### PR TITLE
First demo of learning function

### PR DESCRIPTION
**Pull request by Misty-W**
## Description

Adding tutorial example for learning-based PEC workflow. 

fixes https://github.com/unitaryfund/mitiq/issues/1280

---

### License

- [x] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Before opening the PR, please ensure you have completed the following where appropriate.
- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://google.github.io/styleguide/pyguide.html#383-functions-and-methods) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [x] Added myself / the copyright holder to the AUTHORS file

### CODE REVIEW

**Comment by Misty-W** [comment]
Added a label in `pec.myst` for linking to the PEC section of the user guide (instead of re-explaining PEC in the text of the example).

**Comment by nathanshammah** [comment]
Thanks @Misty-W. Just to clarify, a side comment because there are a couple of PRs open: 
- https://github.com/unitaryfund/mitiq/pull/1395: this PR needs to be updated for the user guide section of the docs, correct? Or shall we close it? I would at least add some information in the PEC users guide under "advanced options" file 4. 
- https://github.com/unitaryfund/mitiq/pull/1510: this is WIP
- https://github.com/unitaryfund/mitiq/pull/1491: this has been put on hold, if I understand correctly

I can provide comments on this PR.

**Comment by Misty-W** [comment]
> Thanks @Misty-W. Just to clarify, a side comment because there are a couple of PRs open:
> 
> * [Write user guide section for learning-based PEC #1395](https://github.com/unitaryfund/mitiq/pull/1395): this PR needs to be updated for the user guide section of the docs, correct? Or shall we close it? I would at least add some information in the PEC users guide under "advanced options" file 4.

Let’s keep the PR for the user guide open and I’ll update it after this PR is closed.

> * [Learn biased noise parameters #1510](https://github.com/unitaryfund/mitiq/pull/1510): this is WIP

Yes, I plan to update this in the next week.

> * [Unmitigated version of learning-based PEC #1491](https://github.com/unitaryfund/mitiq/pull/1491): this has been put on hold, if I understand correctly

Yes, let’s revisit the plan after the other open learning-based PEC PRs are closed.

> I can provide comments on this PR.

Great, thanks! I added you as a reviewer.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:67]:
```suggestion
CNOT_op = next(circuit.findall_operations_with_gate_type(CXPowGate))
moment_index, op, _ = CNOT_op
operations_to_learn = [Circuit(op)]
```
WDYT of this?

I was a little lost when I got down to `noisy_execute` because I didn't remember what `op` and `index` were. I wonder if this code should be moved into the same cell as that one, so it's more fresh in the readers mind.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:74]: Is there anything special about this observable, or was it just chosen at random? If random, can we simplify and go without an explicit observable?

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:99]: This looks to be the default strategy, so I think we should leave it out.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:107]: {func} link is broken. I think the function name needs to start with `.`? I'm not entirely sure.

File: docs/source/conf.py
Comment (natestemen) [inline docs/source/conf.py:336]: It looks like the thumbnail image needs to be moved into `docs/source/_thumbnails` in order to show on the gallery page.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:208]: We should be more explicit here about relative errors. It's great for the final line to show that using this technique decreases your error!

**Review by natestemen — CHANGES_REQUESTED** [review]
Nice work Misty. It's cool to get a better understanding of what you've been working on, in action!

Some general feedback:
- Some more explanation between cells might be helpful instead of describing what the code is doing.
- Inserting some references to learning based PEC would be great! Got to help people sink their teeth in if they want to.

Some technical feedback:
- Some of the lines are really long, which makes it slightly harder to read and give feedback on, especially using GitHub's "suggestion" tool. A line limit on this file might be good, but my favorite way to do it in markdown/markup languages is to stick to a one sentence per line rule.
- The git history on this PR seems to have gotten jumbled, and it's hard to follow how we got to the current state. There are a few commits touching files that are not involved with this PR which is very strange. Happy for you (or I) to rewrite history to clean it up.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:9]: Guess this is an artifact of conversion?

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:16]: Which line?

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:67]: At one point I was using the `next` function here, and I forget why I changed it. It should still work and it looks nicer.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [reply docs/source/examples/learning-depolarizing-noise.myst:16]: line with
```
+++
```

Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:16]: It's a block break. It's used for mapping to cell-based formats, e.g. jupyter.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:66]: Observable has to be defined for `compute_density_matrix`

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Good point!

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [reply docs/source/examples/learning-depolarizing-noise.myst:16]: To validate my understanding, if we had a myst file that contained
```
some text
+++
more text
```
then if you converted this to a jupyter notebook, you would get two distinct markdown cells?

If so, is there a reason we want that here?

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: that fixed it, thanks!

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Agreed. Updated to be relative, but I need to fix the formatting.

File: docs/source/conf.py
Comment (Misty-W) [reply docs/source/conf.py:336]: I thought it was just a rendering issue on my browser, but moving into `docs/source/_thumbnails` was the right call.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:16]: I guess we don't really need it here- I had separate cells for title and first paragraph, and didn't combine them before conversion to `.myst` format. I can take it out in my next round of edits.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Moved to codeblock computing and plotting loss function.

Comment (natestemen) [reply docs/source/examples/learning-depolarizing-noise.myst]: I don't think that's true. The definition of [`compute_density_matrix`](https://github.com/unitaryfund/mitiq/blob/1eca124f764d7b29ef5b751cce729479cdcfb92f/mitiq/interface/mitiq_cirq/cirq_utils.py#L61-L65) doesn't take an observable variable, and in the definition of [`evaluate`](https://github.com/unitaryfund/mitiq/blob/1eca124f764d7b29ef5b751cce729479cdcfb92f/mitiq/executor/executor.py#L119-L125) is the observable variable is optional.

https://github.com/unitaryfund/mitiq/blob/1eca124f764d7b29ef5b751cce729479cdcfb92f/mitiq/interface/mitiq_cirq/cirq_utils.py#L61-L65
https://github.com/unitaryfund/mitiq/blob/1eca124f764d7b29ef5b751cce729479cdcfb92f/mitiq/executor/executor.py#L119-L125

Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Strange, I remember getting an error when setting the `Observable` to `None` when using `evaluate` with `compute_density_matrix`. Anyway, the specified observable happens to work well for the optimization. I've moved it into the code cell for computing the loss function, since I don't think it needs to be emphasized.

**Comment by Misty-W** [comment]
Thanks for the review @natestemen! 
> 
> Some general feedback:
> 
> * Some more explanation between cells might be helpful instead of describing what the code is doing.

Agreed, I've replaced some of the descriptive text with explanations. Let me know what in particular needs additional clarification- it helps to get an outside perspective!

> * Inserting some references to learning based PEC would be great! Got to help people sink their teeth in if they want to.

Thanks for the reminder! I added a citation to the learning-based quantum error mitigation paper, as well as links to the PEC and CDR sections of the guide and the API-doc.

> Some technical feedback:
> 
> * Some of the lines are really long, which makes it slightly harder to read and give feedback on, especially using GitHub's "suggestion" tool. A line limit on this file might be good, but my favorite way to do it in markdown/markup languages is to stick to a one sentence per line rule.

Good point, I've imposed both.

> * The git history on this PR seems to have gotten jumbled, and it's hard to follow how we got to the current state. There are a few commits touching files that are not involved with this PR which is very strange. Happy for you (or I) to rewrite history to clean it up.

Something weird happened to the branch, and it took multiple tries to update it. I didn't think it was possible to rewrite the commit history after pushing to GitHub, but it'd be nice to clean it up.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (andreamari) [inline docs/source/examples/learning-depolarizing-noise.myst:54]: Please check if we need all the imports. E.g. gates

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (andreamari) [inline docs/source/examples/learning-depolarizing-noise.myst:65]: Can we show an example with more than 1 CNOT gate?
You can try to change seed or `n_moments`, or just manually define the circuit without calling the CDR function. 

More than 1 CNOT allows us to test the technique one step further in the direction of a real use case.

If for some reason it doesn't work, we come back to the single CNOT case.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (andreamari) [inline docs/source/examples/learning-depolarizing-noise.myst:75]:
```suggestion
def ideal_execute(circ: Circuit) -> np.ndarray:
    # For large circuits, this should be replaced by a near-Clifford simulator.
    return compute_density_matrix(circ, noise_level=(0.0,))
```

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (andreamari) [inline docs/source/examples/learning-depolarizing-noise.myst:88]: If we go for multiple CNOTs this part may need to change a bit

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (andreamari) [inline docs/source/examples/learning-depolarizing-noise.myst:97]: This need to change if there are more than 1 CNOTs. 
[Circuit.batch_insert()](https://quantumai.google/reference/python/cirq/Circuit#batch_insert) or https://quantumai.google/reference/python/cirq/map_operations_and_unroll may be useful.

**Review by andreamari — COMMENTED** [review]
Thanks @Misty-W, I left some optional suggestions.
Do you have any idea of how long is the execution time?

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:97]: thank you, this is what I was looking for in the Cirq docs and concluded didn't exist!

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Thanks for the reminder! Just pushed a new commit with cleaned-up imports.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Locally this works and takes about 5 min to run. We'll see what it does in CI :)

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Indeed, this is updated to match the new circuit and generalized executor.

**Comment by Misty-W** [comment]
> Thanks @Misty-W, I left some optional suggestions. Do you have any idea of how long is the execution time? Comparing with docs build time on master this seems 15-20 min longer but maybe I am wrong.

Thanks for the suggestions @andreamari! 

Although I've increased the number of moments in the circuit (5 --> 10), I was able to bring the build time down to 25 min (same as latest commit on master [4ea99b4](https://github.com/unitaryfund/mitiq/commit/4ea99b473888a1d828e4f976a383d1ff1ea6eec6)) by reducing the number of points calculated for the loss function plot. 

Anything else that should be addressed before merging?

**Review by andreamari — APPROVED** [review]
LGTM, thanks @Misty-W !
Note: This is our first non-ZNE example!

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:48]: The link to `generate_training_circuits` doesn't link with the API-doc. Should it be `{func}.generate_training_circuits`?

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:99]: Same problem with the link to API-doc with `execute_with_pec`.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [inline docs/source/examples/learning-depolarizing-noise.myst:76]: Casting the generator to list is now unneeded.

**Review by natestemen — APPROVED** [review]
Nice work Misty! I really like this example. It's clear, and shows a nice clear advantage while using this technique.

I've left 3 comments, all of which are optional changes.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:76]: Thanks for asking about this! I will leave it as-is because if I don't cast to list, only one noisy operation is added in the `noisy_executor`.

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Good catch, forgot the `.` before `pec.execute_w_pec`

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst]: Fixed it to be `.cdr.clifford_training_data.generate_training_circuits`

**Comment by Misty-W** [comment]
> Nice work Misty! I really like this example. It's clear, and shows a nice clear advantage while using this technique.
> 
> I've left 3 comments, all of which are optional changes.

Thanks @natestemen !

File: docs/source/examples/learning-depolarizing-noise.myst
Comment (natestemen) [reply docs/source/examples/learning-depolarizing-noise.myst:76]: Fine to leave it out, but are you sure? You later iterate over `CNOT_ops`, rather than just grabbing the first element. [`findall_operations_with_gate_type`](https://github.com/quantumlib/Cirq/blob/946f247dc66f1257b463766536a854bb36b4c3c6/cirq-core/cirq/circuits/circuit.py#L778-L780) returns an iterable so I think it should be fine to leave it as that.

Comment (Misty-W) [reply docs/source/examples/learning-depolarizing-noise.myst:76]: I thought so too, but when I tested it, the loss function plot shifted so that the minimum was around 0.03 instead of the true value of 0.05.  I think that's because only one noisy op was added in the executor if I left CNOT_ops as an iterable.

Comment (natestemen) [reply docs/source/examples/learning-depolarizing-noise.myst:76]: Ohhh okay I see. Yeah we can't iterate over a generator multiple times. Forgot about that! Thanks for helping me get there.

**Pull request merged**
